### PR TITLE
[DEVOPS-1151] Extend addWAddress to support newly-found accounts

### PR DESCRIPTION
**Pull request by adinapoli-iohk**
The previous incarnation of `addWAddress` had a bug: in case of new
accounts (like the ones we could find as part of a restoration) it
wouldn't add the relevant address to the DB, because the old code was
using a lens traversal with 'ix' which notoriously does nothing if the
key (in this case the account index) was not there.

This commit fixes it.

## Type of change

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [ ] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist

- [ ] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [ ] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [ ] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.
- [ ] CHANGELOG entry has been added and is linked to the correct PR on GitHub.

## Testing checklist

- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

## How to merge

Send the message `bors r+` to merge this PR. For more information, see
[`docs/how-to/bors.md`](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/how-to/bors.md).

### CODE REVIEW

**Comment by deepfire** [comment]
Are test failure expected?
```
  test/Test/Pos/Wallet/Web/Mode.hs:247:5:
  1) Test.Pos.Wallet.Web.Methods.Payment.(requiresNetworkMagic=NMMustBeNothing).Wallet.Web.Methods.Payment, Submitting a payment when restoring, should fail with 403
       expected exactly one element, but list is empty (address) (after 1 test):
         WalletTestParams {
           blockTestParams = TestParams {
           start time: 0
           initializer: GenesisInitializer {
           49 poor guys, 6 rich guys , total balance is 25149294507840020, richmen share is 0.563 and useHDAddresses flag is False
           FakeAvvmOptions:
           faoCount: 3
           faoOneBalance: 8

           avvm balance factor: 67874375218776/1000000000000000 (approx. 0.067874375218776)
           heavyDlg: True
           seed: 0
         }

         }

         }

         ([<encrypted key>,<encrypted key>,<encrypted key>,<encrypted key>,<encrypted key>,<encrypted key>,<encrypted key>],[<passphrase>,<passphrase>,<passphrase>,<passphrase>,<passphrase>,<passphrase>,<passphrase>])
         1
         6
         Just (Address {addrRoot = AbstractHash d4750df48a87b06b5787240af551360061d7dae6f81c922c3ea0b910, addrAttributes = Attributes { data: AddrAttributes {aaPkDerivationPath = Just (HDAddressPayload {getHDAddressPayload = "\SUB\162Z\242\152\159\186\241\147Jq\SO[\RS8{\166\153\t\223q\NAKmF\164\190\227:"}), aaStakeDistribution = BootstrapEraDistr, aaNetworkMagic = NMNothing} }, addrType = ATPubKey})
         3

  test/Test/Pos/Wallet/Web/Mode.hs:247:5:
  2) Test.Pos.Wallet.Web.Methods.Payment.(requiresNetworkMagic=NMMustBeNothing).Wallet.Web.Methods.Payment, One payment, Send money from one own address to multiple own addresses; check balances validity for destination addresses, source address and change address; validate history and used/change addresses
       expected exactly one element, but list is empty (address) (after 1 test):
         WalletTestParams {
           blockTestParams = TestParams {
           start time: 0
           initializer: GenesisInitializer {
           49 poor guys, 6 rich guys , total balance is 25149294507840020, richmen share is 0.563 and useHDAddresses flag is False
           FakeAvvmOptions:
           faoCount: 3
           faoOneBalance: 8

           avvm balance factor: 67874375218776/1000000000000000 (approx. 0.067874375218776)
           heavyDlg: True
           seed: 0
         }

         }

         }

         ([<encrypted key>,<encrypted key>,<encrypted key>,<encrypted key>,<encrypted key>,<encrypted key>,<encrypted key>],[<passphrase>,<passphrase>,<passphrase>,<passphrase>,<passphrase>,<passphrase>,<passphrase>])
         1
         6
         Just (Address {addrRoot = AbstractHash d4750df48a87b06b5787240af551360061d7dae6f81c922c3ea0b910, addrAttributes = Attributes { data: AddrAttributes {aaPkDerivationPath = Just (HDAddressPayload {getHDAddressPayload = "\SUB\162Z\242\152\159\186\241\147Jq\SO[\RS8{\166\153\t\223q\NAKmF\164\190\227:"}), aaStakeDistribution = BootstrapEraDistr, aaNetworkMagic = NMNothing} }, addrType = ATPubKey})
         3

  test/Test/Pos/Wallet/Web/Mode.hs:247:5:
  3) Test.Pos.Wallet.Web.Methods.Payment.(requiresNetworkMagic=NMMustBeJust).Wallet.Web.Methods.Payment, Submitting a payment when restoring, should fail with 403
       expected exactly one element, but list is empty (address) (after 1 test):
         WalletTestParams {
           blockTestParams = TestParams {
           start time: 0
           initializer: GenesisInitializer {
           49 poor guys, 6 rich guys , total balance is 25149294507840020, richmen share is 0.563 and useHDAddresses flag is False
           FakeAvvmOptions:
           faoCount: 3
           faoOneBalance: 8

           avvm balance factor: 67874375218776/1000000000000000 (approx. 0.067874375218776)
           heavyDlg: True
           seed: 0
         }

         }

         }

         ([<encrypted key>,<encrypted key>,<encrypted key>,<encrypted key>,<encrypted key>,<encrypted key>,<encrypted key>],[<passphrase>,<passphrase>,<passphrase>,<passphrase>,<passphrase>,<passphrase>,<passphrase>])
         1
         6
         Just (Address {addrRoot = AbstractHash f652f78e7a89e5240e6ab999d4ada6bd6fa9429732f80b30805c453e, addrAttributes = Attributes { data: AddrAttributes {aaPkDerivationPath = Just (HDAddressPayload {getHDAddressPayload = "\SUB\162Z\242\152\159\186\241\147Jq\SO[\RS8{\166\153\t\223q\NAKmF\164\190\227:"}), aaStakeDistribution = BootstrapEraDistr, aaNetworkMagic = NMJust 1017530110} }, addrType = ATPubKey})
         3

  test/Test/Pos/Wallet/Web/Mode.hs:247:5:
  4) Test.Pos.Wallet.Web.Methods.Payment.(requiresNetworkMagic=NMMustBeJust).Wallet.Web.Methods.Payment, One payment, Send money from one own address to multiple own addresses; check balances validity for destination addresses, source address and change address; validate history and used/change addresses
       expected exactly one element, but list is empty (address) (after 1 test):
         WalletTestParams {
           blockTestParams = TestParams {
           start time: 0
           initializer: GenesisInitializer {
           49 poor guys, 6 rich guys , total balance is 25149294507840020, richmen share is 0.563 and useHDAddresses flag is False
           FakeAvvmOptions:
           faoCount: 3
           faoOneBalance: 8

           avvm balance factor: 67874375218776/1000000000000000 (approx. 0.067874375218776)
           heavyDlg: True
           seed: 0
         }

         }

         }

         ([<encrypted key>,<encrypted key>,<encrypted key>,<encrypted key>,<encrypted key>,<encrypted key>,<encrypted key>],[<passphrase>,<passphrase>,<passphrase>,<passphrase>,<passphrase>,<passphrase>,<passphrase>])
         1
         6
         Just (Address {addrRoot = AbstractHash f652f78e7a89e5240e6ab999d4ada6bd6fa9429732f80b30805c453e, addrAttributes = Attributes { data: AddrAttributes {aaPkDerivationPath = Just (HDAddressPayload {getHDAddressPayload = "\SUB\162Z\242\152\159\186\241\147Jq\SO[\RS8{\166\153\t\223q\NAKmF\164\190\227:"}), aaStakeDistribution = BootstrapEraDistr, aaNetworkMagic = NMJust 1017530110} }, addrType = ATPubKey})
         3

Randomized with seed 1912705277

Finished in 49.5702 seconds
31 examples, 4 failures, 6 pending
```

File: wallet/src/Pos/Wallet/Web/State/Storage.hs [hunk 1]
```diff
@@ -615,21 +616,29 @@ createWallet cWalId cWalMeta isReady curTime = do
...
+    where
+        freshAccount :: AccountInfo
+        freshAccount = AccountInfo (WebTypes.CAccountMeta "New account") mempty mempty 0
+
+        aux :: Maybe AccountInfo -> Maybe AccountInfo
+        aux Nothing     = Just $ addAddress freshAccount
+        aux (Just info) = Just $ addAddress info
+
+        addAddress :: AccountInfo -> AccountInfo
+        addAddress info@AccountInfo{_aiUnusedKey, _aiAddresses}
```
[the remark this answers is not in the document]
Comment (adinapoli-iohk) [reply wallet/src/Pos/Wallet/Web/State/Storage.hs]: I think you might be right -- probably the correct way to go is, when creating a `freshAccount`, to also pre-populate it with the newly-found address before handling the case where the account is already there and has a bunch of addresses in it.

**Comment by tatyanavych** [comment]
1.3.2 has been released. Possibly we need to have it in the next hotfix 1.3.3.

**Comment by tatyanavych** [comment]
1.3.2 has been released. Possibly we need to have this commit in the next cardano-sl hotfix 1.3.3.

**Comment by KtorZ** [comment]
bors +r

**Comment by rvl** [comment]
The `release/1.3.3` branch doesn't have `bors.toml`, so bors won't work here.

**Pull request merged**